프로젝트 생성 및 상품 목록 화면 구성

hyeffie opened the pull request:
## 🧩 주요 작업

프로젝트를 생성하고 필요한 리소스를 에셋으로 추가했습니다.

- [x]  프로젝트 생성
- [x]  ColorAsset 추가

1단계 요구사항인 상품 목록 화면을 구성했습니다.

- [x]  Home 화면을 스토리보드로 구성
- [x]  상품 종류를 나타내는 ProductCategory 타입 추가
- [x]  상품 리스트 collection view cell xib 파일 추가
- [x]  ProductCollectionViewCell 구현
- [x]  이벤트 태그 label 추가
- [x]  리스트 collection view 헤더 xib 파일 추가
- [x]  ProductCollectionHeader 구현
- [x]  HomeViewController 구현

## 🧩 스크린샷

![Simulator Screen Recording - iPhone 14 Pro - 2023-04-19 at 18 38 32](https://user-images.githubusercontent.com/56967908/233039016-df06433c-bdfc-4cb6-83d9-ea4d9ac8b5c9.gif)


## 🧩 기술 키워드

**구현 과정에서 학습하거나 알게 된 키워드**

- xib
- UICollectionView
- UICollectionViewCell
- UICollectionReusableView
- NSMutableAttributedString

## 🧩 고민과 해결 과정

### cell reusable 이슈

cell을 만들 때 configure로 데이터를 전달해서 태그가 있을 때 stack view에 태그 label을 추가하도록 구현했는데, dequeueReusableCell로 셀을 재사용하면서 stack에 tag가 이어서 추가되는 문제가 있었습니다.

재사용 시점에 stack 에 담겨 있던 태그들을 제거하고 다시 태그를 추가해주고자 prepareForReuse에서 stackview의 subviews를 iterate 하면서 super로 부터 제거하는 코드를 작성했는데 다음 스크린샷처럼 태그 label이 제대로 사라지지 않다는 걸 확인했습니다.

![viewdebug](https://user-images.githubusercontent.com/56967908/233039159-d0342d33-99c0-4284-8a48-d006f287427a.png)


검색 후에 stack에서 element를 관리하려면 subviews가 아닌 arrangedSubviews를 사용해야 한다는 것을 알게 되었고, 정상 동작하도록 수정했습니다.

hyeffie committed:
Init Xcode Project

- iOS 프로젝트 생성

hyeffie committed:
Add color assets

hyeffie committed:
Add product list

- 상품 목록 화면 구성
- HomeViewController 구현
- 프로젝트 내부 파일 폴더 정리

hyeffie committed:
Hide tag stack view when there is no tag

- 태그가 없을 때 화면 대응
- 상품 정보 뷰 레이아웃 재구성

godrm commented on Onban03/Onban03/Home/ProductCell/ProductCollectionViewCell.swift:
```diff
@@ -0,0 +1,59 @@
+    super.layoutSubviews()
+    
+    thumbnailImageView.layer.cornerRadius = 5
+    thumbnailImageView.layer.masksToBounds = true
```
이 정도 속성 변경은 여기보다 awakeFromNib 에서 처리하는 게 좋습니다
그 이유는 여기는 계속 반복해서 호출되는 곳이고, 아래는 로딩하고 나서 호출되는 곳이라서 그렇습니다.

godrm commented on Onban03/Onban03/Home/ProductCell/ProductCollectionViewCell.swift:
```diff
@@ -0,0 +1,59 @@
+      return
+    }
+    
+    for (i, tag) in tags.enumerated() {
```
i처럼 너무 축약된 표현보다 index면 그 의미를 잘 드러내는 단어를 그대로 쓰는 게 읽기 편할 것 같습니다

godrm commented on Onban03/Onban03/Home/ProductCell/TagLabel.swift:
```diff
@@ -0,0 +1,44 @@
+    text = name
+    font = UIFont.systemFont(ofSize: 8, weight: .bold, width: .standard)
+    textColor = .white
+    setBackgroundColor(index)
```
뭔가 설정하는 의미의 메소드라면 생성자를 활용하는 것은 어떨까요?

godrm commented on Onban03/Onban03/Home/HomeViewController.swift:
```diff
@@ -0,0 +1,73 @@
+  
+  func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+    let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "productCell", for: indexPath) as! ProductCollectionViewCell
+    cell.configure(tags: [Tag(name: "BEST"), Tag(name: "NEW"), Tag(name: "이벤트특가")])
```
Tag 배열이 하드코딩되어 있는 것은 중복이나 수정하기 어려운 코드가 됩니다. 
이런 값도 보관하는 모델을 설계하려고 노력해보세요
콜랙션 뷰의 데이터 소스를 갖고 있는 모델이 필요할 것 같습니다.

godrm commented on Onban03/Onban03/Home/Header/ProductCollectionHeader.swift:
```diff
@@ -0,0 +1,25 @@
+  
+  override init(frame: CGRect) {
+    super.init(frame: frame)
+  }
```
아까 추가했던 지정 생성자인데 필요없으면 지우셔도 됩니다.

godrm reviewed (approved):
참고로 이번 미션부터는 1단계-2단계 이런 식으로 나눠져 있는 게 아닙니다. 
전체가 요구사항이고 화면에 대한 요구사항이 먼저 설명되어 있을 뿐입니다. 
전체에 대한 계획을 팀에서 알아서 하루 단위로 나눠서 계획하고 구현하시면 됩니다. 
수고하셨습니다. 머지하고 진행하셔도 될 것 같습니다.

hyeffie committed:
Relocation implementation (layoutSubviews -> awakeFromNib)

- 재사용과 관련 없는 thumbnailImageView 속성 변경 코드를 awakeFromNib으로 이동

hyeffie committed:
Rename addTags iteration constant (i -> index)

- addTags 메소드 내 반복 상수명 변경

hyeffie committed:
Add convenience initializer to TagLabel

TagLabel에 index와 name을 받는 간편 생성자 추가
설정 관련 구현을 메소드에서 생성자로 이동

hyeffie committed:
Remove useless required designated initializers

- frame과 coder를 전달받는 불필요 지정 생성자 삭제

hyeffie committed:
Add Product

- Product 구조체를 추가하고 모델 그룹핑

wood0203 commented on Onban03/Onban03/Home/ProductCell/ProductCollectionViewCell.swift:
```diff
@@ -0,0 +1,59 @@
+
+class ProductCollectionViewCell: UICollectionViewCell {
+  
+  @IBOutlet weak var thumbnailImageView: UIImageView!
```
썸네일 작명 센스가 너무 좋네요 눈에 잘들어와요
이름에 -ImageView는 삭제하셔도 좋을거 같습니다

hyeffie replied:
오 이 부분 일관성 지킬 수 있을지 한 번 고민해봐야겠네요. 감사합니다 !

wood0203 commented on Onban03/Onban03/Home/ProductCell/ProductCollectionViewCell.swift:
```diff
@@ -0,0 +1,59 @@
+  
+  func addTags(_ tags: [Tag]) {
+    guard tags.isEmpty == false else {
+      tagStackView.superview?.isHidden = true
```
태그가 없을 때 어떡할지 고민하고 있었는데 한 수 배워가요

wood0203 commented on Onban03/Onban03/Home/ProductCell/TagLabel.swift:
```diff
@@ -0,0 +1,44 @@
+  }
+  
+  func configure(index: Int, name: String) {
+    text = name
+    font = UIFont.systemFont(ofSize: 8, weight: .bold, width: .standard)
+    textColor = .white
```
해당 변수들이 선언되지 않고 사용되지 않고 있는거 같은데 
혹시 제가 이해를 못한거라면 설명좀 부탁드려도 될까요?

hyeffie replied:
self 가 누락돼서 오해할 수 있겠네요 🥲 self 관련 스타일을 재고해보겠습니다.. 감사해요!

wood0203 reviewed (commented):
코드 보고 많은걸 깨닫고 배움이 늘었읍니다..^^

hyeffie committed:
Add DataMock

- 더미데이터 추가
- collection view 에 표시할 데이터를 section별로 나누기 위한 CategoryProducts 구조체 추가
- Product 일부 속성 옵셔널화
- cell configure 메소드 구현 추가
- header configure 메소드 구현 추가
- collection view data source 메소드 구현 수정
- Tag 구조체 파일 생성

hyeffie committed:
Add Int won format extension

- 원화 형식 문자열을 반환하는 wonFormat 메소드 추가
- 관련 구현 수정

hyeffie committed:
Apply Coding Style Guideline

- 코딩 스타일 가이드라인 적용

hyeffie commented:
@godrm 남겨주신 코멘트 모두 확인하고 반영했습니다! 감사합니다 🙏(에디) 😃(에피)

hyeffie commented:
@dpfdlalfm 우드 리뷰 모두 확인했습니다. 꼼꼼하게 리뷰 해주셔서 감사합니다 😊
